### PR TITLE
IoTV2: Introduce pipeTaskRestartTimes for ensuring replicate events as order

### PR DESCRIPTION
**Pull request by Pengzna**
as title

### CODE REVIEW

**Comment by Pengzna** [comment]
@OneSizeFitsQuorum @Caideyipi @133tosakarin PTAL

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java [hunk 1]
```diff
@@ -1385,7 +1404,7 @@ private TPipeConsensusTransferResp onRequest(
             // when the last seal req is applied, we can discard this event.
             if (resp != null
                 && resp.getStatus().getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
-              onSuccess(onSyncedCommitIndex + 1, isTransferTsFileSeal);
+              onSuccess(tCommitId, isTransferTsFileSeal);
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java:1407]: It looks like there's a +1 before there, is there any inconsistency?

Comment (Pengzna) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java:1407]: there is not inconsistency. tCommitId.getCommitIndex is always equal to onSyncedCommitIndex + 1 here

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java [hunk 2]
```diff
@@ -1463,10 +1482,26 @@ private void resetWithNewestRebootTime(int connectorRebootTimes) {
...
+    private void resetWithNewestRestartTime(int pipeTaskRestartTimes) {
+      LOGGER.info(
+          "PipeConsensus-PipeName-{}: receiver detected an newer pipeTaskRestartTimes, which indicates the pipe task has restarted. receiver will reset all its data.",
+          consensusPipeName);
+      // since pipe task will resend all data that hasn't synchronized, it's safe to clear all
+      // events in buffer.
+      this.reqExecutionOrderBuffer.clear();
+      this.onSyncedCommitIndex = 0;
+      this.pipeTaskRestartTimes = pipeTaskRestartTimes;
       this.tsFileWriterPool.handleExit(consensusPipeName);
```
Comment (133tosakarin) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java:1505]: How about writing it like this?

`    private void resetRebootTime(int rebootTime) {
      this.connectorRebootTimes = rebootTime;
      clear();
    }

    private void resetPipeRestartTime(int restartTime) {
      this.pipeTaskRestartTime = restartTime;
      clear();
    }

    private void clear() {
      this.reqExecutionOrderBuffer.clear();
      this.tsFileWriterPool.handleExit(consensusPipeName);
      this.onSyncedCommitIndex = 0;
    }`

Comment (Pengzna) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/receiver/protocol/pipeconsensus/PipeConsensusReceiver.java:1505]: fixed

**Review by 133tosakarin — APPROVED** [review]

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**